### PR TITLE
feat: Check where clauses when searching for trait impls

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Summary\*

Previously when searching for a trait impl for a type such as `[A]`, an impl such as `impl Foo for [T] where T: Eq` would be selected without verifying whether `A` implemented `Eq` at all. This would later cause a panic during monomorphization if there was no such impl. For this PR, I've added a check to recursively verify all trait constraints have impls.

## Additional Context

The error that is issued in the example program is:

```
error: No matching impl found for `Field: Eq`
  ┌─ /../program/src/main.nr:4:12
  │
4 │     assert(a.eq(a));
  │            ------- No impl for `Field: Eq`
  │
  = Required by `[Field; 2]: Eq`
  = Required by `[[Field; 2]; 2]: Eq`
  = Required by `[[[Field; 2]; 2]; 2]: Eq`
  = Required by `[[[[Field; 2]; 2]; 2]; 2]: Eq`
```

Note the "Required by" clauses which are helpful in tracking down where the constraint originated from. This is vital to cases like the above where the failing `Field: Eq` impl is not a result of `a` being a `FIeld` directly, it is because `a` is a `[[[[Field; 2]; 2]; 2]; 2]` which indirectly requires `Field: Eq`.

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by jfecher** [comment]
I've temporarily disabled the `impl_with_where_clause` test. It has just started being affected by an existing issue related to trait dispatch: https://github.com/noir-lang/noir/issues/3409. Since the issue is existing and I already plan to fix it in another PR, I figured I will leave that out of this PR since I expect it may be a larger fix.

File: tooling/nargo_cli/tests/compile_success_empty/impl_with_where_clause/src/main.nr [hunk 1]
```diff
@@ -1,18 +1,19 @@
 
 fn main() {
-    let array: [Field; 3] = [1, 2, 3];
-    assert(array.eq(array));
+    // Test is temporarily disabled
```
Comment (vezenovm) [inline tooling/nargo_cli/tests/compile_success_empty/impl_with_where_clause/src/main.nr:3]: Do we have a followup issue for this? Looks to be #3409? Can we link it here just so we do not forget.

**Comment by kevaundray** [comment]
`compile_success_empty_trait_where_clause` is still failing, instead of commenting out parts of the Noir program, can you put it into the compile_failure folder so that when it is fixed the CI will fail, reminding us to move it back?

**Review by vezenovm — COMMENTED** [review]
Changes look good. I think you missed disabling one of the compilation tests for traits as well. Once that is fixed I can approve.

**Review by kevaundray — APPROVED** [review]
Left comment about `compile_success_empty_trait_where_clause`

File: tooling/nargo_cli/tests/compile_success_empty/impl_with_where_clause/src/main.nr (hunk 1 above)
Comment (jfecher) [reply tooling/nargo_cli/tests/compile_success_empty/impl_with_where_clause/src/main.nr:3]: Yes, I just made #3409 for it

**Comment by jfecher** [comment]
> compile_success_empty_trait_where_clause is still failing

That is odd, it isn't failing locally for me. With it commented out, there should be no reason to do an impl search.

**Comment by jfecher** [comment]
>  can you put it into the compile_failure folder so that when it is fixed the CI will fail, reminding us to move it back?

No, since it is crashing and not just issuing a compiler error, it will still be read as a CI failure.

**Comment by jfecher** [comment]
It turns out the failing test was just a different test with a similar name, so 2 tests are failing now which I've disabled.

**Comment by kevaundray** [comment]
> > can you put it into the compile_failure folder so that when it is fixed the CI will fail, reminding us to move it back?
> 
> No, since it is crashing and not just issuing a compiler error, it will still be read as a CI failure.

Oh didn't notice we changed to distinguish between panics and compile failures

**Comment by Savio-Sou** [comment]
Does this require any documentation?

**Comment by jfecher** [comment]
@Savio-Sou No, sorry. I'm waiting to document traits until they're stabilized. Until then I don't want to advertise to users that they're technically available but are unstable and buggy.

**Pull request merged**